### PR TITLE
rune/libenclave/intelsgx && attestation: both support quote version 2 and version 3

### PR DESCRIPTION
**Pull request by hustliyilin**
Signed-off-by: Yilin Li <YiLin.Li@linux.alibaba.com>

### CODE REVIEW

**Comment by hustliyilin** [comment]
This is a draft. @jiazhang0

File: rune/libenclave/attestation/sgx/ias/ias.go [hunk 1]
```diff
@@ -106,21 +106,20 @@ func NewIasAttestation(cfg map[string]string) (*IasAttestation, error) {
...
+		quote.AttestationKeyType)
+	logrus.Debugf("  Quoting Enclave SVN:                  			  %d\n",
+		quote.QeSvn)
+	logrus.Debugf("  PCE SVN:                              			  %d\n",
+		quote.PceSvn)
+	logrus.Debugf("  Quoting Enclave Vendor Id:            			  0x%v\n",
+		hex.EncodeToString(quote.QeVendorId[:]))
+	logrus.Debugf("  User Data:                            			  0x%v\n",
+		hex.EncodeToString(quote.UserData[:]))
+	logrus.Debugf("  Report:                               			  ...\n")
```
Comment (jiazhang0) [inline rune/libenclave/attestation/sgx/ias/ias.go:122]: Why you change the quote's structure in IAS specific code?

**Review by jiazhang0 — CHANGES_REQUESTED** [review]

File: rune/libenclave/attestation/sgx/ias/ias.go [hunk 2]
```diff
@@ -106,57 +108,33 @@ func NewIasAttestation(cfg map[string]string) (*IasAttestation, error) {
...
-		hex.EncodeToString(q[intelsgx.QuoteLength:intelsgx.QuoteLength+32]))
-
-	if quote.Version != intelsgx.QuoteVersion {
+	err := intelsgx.DumpQuote(q)
+	if err != nil {
+		return err
+	}
+
+	if quote.Version != intelsgx.QuoteVersion2 && quote.Version != intelsgx.QuoteVersion3 {
 		return fmt.Errorf("Invalid quote version: %d", quote.Version)
```
Comment (jiazhang0) [inline rune/libenclave/attestation/sgx/ias/ias.go:117]: Change this message to "Unsupported quote version".

File: rune/libenclave/attestation/sgx/ias/ias.go [hunk 3]
```diff
@@ -106,57 +108,33 @@ func NewIasAttestation(cfg map[string]string) (*IasAttestation, error) {
...
 	}
 
-	if quote.SignatureType != intelsgx.QuoteSignatureTypeUnlinkable &&
-		quote.SignatureType != intelsgx.QuoteSignatureTypeLinkable {
-		return fmt.Errorf("Invalid signature type: %#04x", quote.SignatureType)
+	if quote.AttestationKeyType != intelsgx.QuoteSignatureTypeUnlinkable &&
+		quote.AttestationKeyType != intelsgx.QuoteSignatureTypeLinkable {
+		return fmt.Errorf("Invalid attestation key type: %#04x", quote.AttestationKeyType)
```
Comment (jiazhang0) [inline rune/libenclave/attestation/sgx/ias/ias.go:122]: Change to "Unsupported attestation key type".

File: rune/libenclave/intelsgx/aesmd.go [hunk 4]
```diff
@@ -135,6 +135,59 @@ func DumpLocalReport(report []byte) error {
...
+
+		logrus.Debugf("  Quoting Enclave SVN:                  %d\n",
+			quoteBody.QeSvn)
+		logrus.Debugf("  PCE SVN:                              %d\n",
+			quoteBody.PceSvn)
+		logrus.Debugf("  Quoting Enclave Vendor Id:            0x%v\n",
+			hex.EncodeToString(quoteBody.QeVendorId[:]))
+		logrus.Debugf("  User Data:                            0x%v\n",
+			hex.EncodeToString(quoteBody.UserData[:]))
+
```
Comment (jiazhang0) [inline rune/libenclave/intelsgx/aesmd.go:178]: plz remove this empty line.

File: rune/libenclave/intelsgx/aesmd.go [hunk 5]
```diff
@@ -135,6 +135,59 @@ func DumpLocalReport(report []byte) error {
...
+	} else {
+		return fmt.Errorf("Unsupported Quote Version: %d", q.Version)
+	}
+
+	logrus.Debugf("  Report:                               ...\n")
```
Comment (jiazhang0) [inline rune/libenclave/intelsgx/aesmd.go:183]: If you decide to dump the report, plz remove this line.

Comment (jiazhang0) [reply rune/libenclave/intelsgx/aesmd.go:183]: However, the dump messages for local report have a wrong indent. So dump function should have a parameter for indent size to get a good result view.

File: rune/libenclave/intelsgx/arch.go [hunk 6]
```diff
@@ -128,26 +129,29 @@ type Report struct {
 }
 
 type ReportBody struct {
-	CpuSvn       [16]byte `struct:"[16]byte"`
-	MiscSelect   uint32   `struct:"uint32"`
-	_            [12]byte `struct:"[12]byte"`
+	CpuSvn     [16]byte `struct:"[16]byte"`
+	MiscSelect uint32   `struct:"uint32"`
+	_          [12]byte `struct:"[12]byte"`
+	// IsvExtProdId is reserved in Quote V3
```
Comment (jiazhang0) [inline rune/libenclave/intelsgx/arch.go:135]: The quote spec for ecdsa has a wrong format definition, plz ignore it. Just remove this comment line and keep the previous code.

File: rune/libenclave/intelsgx/arch.go [hunk 7]
```diff
@@ -128,26 +129,29 @@ type Report struct {
...
 	MrSigner     [32]byte `struct:"[32]byte"`
 	_            [32]byte `struct:"[32]byte"`
-	ConfigId     [64]byte `struct:"[64]byte"`
-	IsvProdId    uint16   `struct:"uint16"`
-	IsvSvn       uint16   `struct:"uint16"`
-	ConfigSvn    uint16   `struct:"uint16"`
-	_            [42]byte `struct:"[42]byte"`
-	IsvFamilyId  [16]byte `struct:"[16]byte"`
-	ReportData   [64]byte `struct:"[64]byte"`
+	// ConfigId is reserved in Quote V3
```
Comment (jiazhang0) [inline rune/libenclave/intelsgx/arch.go:142]: ditto

**Review by jiazhang0 — CHANGES_REQUESTED** [review]

File: rune/libenclave/intelsgx/aesmd.go (hunk 5 above)
Comment (hustliyilin) [reply rune/libenclave/intelsgx/aesmd.go]: > However, the dump messages for local report have a wrong indent. So dump function should have a parameter for indent size to get a good result view.

This issue is not solved now, I will add it as soon as possible @jiazhang0 .

File: rune/libenclave/intelsgx/aesmd.go [hunk 8]
```diff
@@ -100,9 +100,12 @@ func DumpTargetInfo(targetInfo []byte) error {
 	return nil
 }
 
-func DumpLocalReport(report []byte) error {
+/* The report can be used as a separate structure or as a member of the quote structure,
+ *  so we add the indent paramter to both support to dump these two case
+ */
+func DumpLocalReport(report []byte, indent uint32) error {
```
Comment (hustliyilin) [inline rune/libenclave/intelsgx/aesmd.go:106]: I add the indent parameters and the corresponding comments here for DumpLocalReport @jiazhang0 . Thanks. I revise my codes already according to your suggestions.

File: rune/libenclave/intelsgx/aesmd.go [hunk 9]
```diff
@@ -100,9 +100,12 @@ func DumpTargetInfo(targetInfo []byte) error {
 	return nil
 }
 
-func DumpLocalReport(report []byte) error {
+/* The report can be used as a separate structure or as a member of the quote structure,
+ *  so we add the indent paramter to both support to dump these two cases
```
Comment (jiazhang0) [inline rune/libenclave/intelsgx/aesmd.go:104]: Fix this comment line.

**Review by jiazhang0 — CHANGES_REQUESTED** [review]

File: rune/libenclave/attestation/sgx/ias/ias.go (hunk 1 above)
Comment (zhiminghufighting) [reply rune/libenclave/attestation/sgx/ias/ias.go]: Quote structure should be defined in a general package and attester/challenger/verifier can be imported

File: rune/libenclave/attestation/sgx/ias/ias.go [hunk 10]
```diff
@@ -106,57 +108,33 @@ func NewIasAttestation(cfg map[string]string) (*IasAttestation, error) {
...
-		return fmt.Errorf("Invalid quote version: %d", quote.Version)
-	}
-
-	if quote.SignatureType != intelsgx.QuoteSignatureTypeUnlinkable &&
-		quote.SignatureType != intelsgx.QuoteSignatureTypeLinkable {
-		return fmt.Errorf("Invalid signature type: %#04x", quote.SignatureType)
+	err := intelsgx.DumpQuote(q)
+	if err != nil {
+		return err
+	}
```
[the remark this answers is not in the document]
Comment (hustliyilin) [reply rune/libenclave/attestation/sgx/ias/ias.go:112]: > Move this chunk of code ahead of `quote := (*intelsgx.Quote)(unsafe.Pointer(&q[0]))`.

Done, thanks.

File: rune/libenclave/intelsgx/arch.go (hunk 7 above)
Comment (zhiminghufighting) [reply rune/libenclave/intelsgx/arch.go]: Please add "signature[]" into quote structure as below:
typedef struct _quote_t
{
    uint16_t            version;        /* 0   */
    uint16_t            sign_type;      /* 2   */
    sgx_epid_group_id_t epid_group_id;  /* 4   */
    sgx_isv_svn_t       qe_svn;         /* 8   */
    sgx_isv_svn_t       pce_svn;        /* 10  */
    uint32_t            xeid;           /* 12  */
    sgx_basename_t      basename;       /* 16  */
    sgx_report_body_t   report_body;    /* 48  */
    uint32_t            signature_len;  /* 432 */
    uint8_t             signature[];    /* 436 */
} sgx_quote_t;


typedef struct _sgx_quote3_t {
    sgx_quote_header_t  header;              ///< 0:   The quote header.
    sgx_report_body_t   report_body;         ///< 48: The REPORT of the app that is attesting remotely.
    uint32_t            signature_data_len;  ///< 432: The length of the signature_data.  Varies depending on the type of sign_type.
#ifdef _MSC_VER
#pragma warning(push)
#pragma warning ( disable:4200 )
#endif
    uint8_t             signature_data[];    ///< 436: Contains the variable length containing the quote signature and support data for the signature.
#ifdef _MSC_VER
#pragma warning(pop)
#endif
} sgx_quote3_t;


You can see whatever quote or quotes, they have the same content in the first 436 bytes.

Comment (zhiminghufighting) [reply rune/libenclave/intelsgx/arch.go]: Please see two quote header, you can define them in as sgx_quote_header_t and sgx_quote3_header_t
quote header：
{
uint16_t version; /* 0 /
uint16_t sign_type; / 2 /
sgx_epid_group_id_t epid_group_id; / 4 /
sgx_isv_svn_t qe_svn; / 8 /
sgx_isv_svn_t pce_svn; / 10 /
uint32_t xeid; / 12 /
sgx_basename_t basename; / 16 /
}
quote3 header：
typedef struct _sgx_quote3_header_t {
    uint16_t            version;             ///< 0:  The version this quote structure.
    uint16_t            att_key_type;        ///< 2:  sgx_attestation_algorithm_id_t.  Describes the type of signature in the signature_data[] field.
    uint32_t            att_key_data_0;      ///< 4:  Optionally stores additional data associated with the att_key_type.
    sgx_isv_svn_t       qe_svn;              ///< 8:  The ISV_SVN of the Quoting Enclave when the quote was generated.
    sgx_isv_svn_t       pce_svn;             ///< 10: The ISV_SVN of the PCE when the quote was generated.
    uint8_t             vendor_id[16];       ///< 12: Unique identifier of QE Vendor.
    uint8_t             user_data[20];       ///< 28: Custom attestation key owner data.
} sgx_quote3_header_t;

Comment (hustliyilin) [reply rune/libenclave/intelsgx/arch.go]: > Please see two quote header, you can define them in as sgx_quote_header_t and sgx_quote3_header_t
> quote header：
> {
> uint16_t version; /* 0 /
> uint16_t sign_type; / 2 /
> sgx_epid_group_id_t epid_group_id; / 4 /
> sgx_isv_svn_t qe_svn; / 8 /
> sgx_isv_svn_t pce_svn; / 10 /
> uint32_t xeid; / 12 /
> sgx_basename_t basename; / 16 /
> }
> quote3 header：
> typedef struct _sgx_quote3_header_t {
> uint16_t version; ///< 0: The version this quote structure.
> uint16_t att_key_type; ///< 2: sgx_attestation_algorithm_id_t. Describes the type of signature in the signature_data[] field.
> uint32_t att_key_data_0; ///< 4: Optionally stores additional data associated with the att_key_type.
> sgx_isv_svn_t qe_svn; ///< 8: The ISV_SVN of the Quoting Enclave when the quote was generated.
> sgx_isv_svn_t pce_svn; ///< 10: The ISV_SVN of the PCE when the quote was generated.
> uint8_t vendor_id[16]; ///< 12: Unique identifier of QE Vendor.
> uint8_t user_data[20]; ///< 28: Custom attestation key owner data.
> } sgx_quote3_header_t;

Sure, I already achieve this in `arch.go` in the pr.

Comment (hustliyilin) [reply rune/libenclave/intelsgx/arch.go]: > Please add "signature[]" into quote structure as below:
> typedef struct _quote_t
> {
> uint16_t version; /* 0 _/ uint16_t sign_type; /_ 2 _/ sgx_epid_group_id_t epid_group_id; /_ 4 _/ sgx_isv_svn_t qe_svn; /_ 8 _/ sgx_isv_svn_t pce_svn; /_ 10 _/ uint32_t xeid; /_ 12 _/ sgx_basename_t basename; /_ 16 _/ sgx_report_body_t report_body; /_ 48 _/ uint32_t signature_len; /_ 432 _/ uint8_t signature[]; /_ 436 */
> } sgx_quote_t;
> 
> typedef struct _sgx_quote3_t {
> sgx_quote_header_t header; ///< 0: The quote header.
> sgx_report_body_t report_body; ///< 48: The REPORT of the app that is attesting remotely.
> uint32_t signature_data_len; ///< 432: The length of the signature_data. Varies depending on the type of sign_type.
> #ifdef _MSC_VER
> #pragma warning(push)
> #pragma warning ( disable:4200 )
> #endif
> uint8_t signature_data[]; ///< 436: Contains the variable length containing the quote signature and support data for the signature.
> #ifdef _MSC_VER
> #pragma warning(pop)
> #endif
> } sgx_quote3_t;
> 
> You can see whatever quote or quote3, they have the same content in the first 436 bytes.

Sure, I already achieve this in `arch.go` in the pr.

File: rune/libenclave/intelsgx/arch.go [hunk 11]
```diff
@@ -159,17 +161,30 @@ const (
...
-	Version       uint16   `struct:"uint16"`
-	SignatureType uint16   `struct:"uint16"`
-	Gid           uint32   `struct:"uint32"`
-	ISVSvnQe      uint16   `struct:"uint16"`
-	ISVSvnPce     uint16   `struct:"uint16"`
-	_             [4]byte  `struct:"[4]byte"`
-	Basename      [32]byte `struct:"[32]byte"`
+type QuoteHeader struct {
+	Version            uint16 `struct:"uint16"`
+	AttestationKeyType uint16 `struct:"uint16"`
```
Comment (zhiminghufighting) [inline rune/libenclave/intelsgx/arch.go:170]: ‘AttestationKeyType‘ is used by quote3_header (uint16_t att_key_type; ///< 2: sgx_attestation_algorithm_id_t. ), could you set this section name more neutral name?

Comment (hustliyilin) [reply rune/libenclave/intelsgx/arch.go:170]: > ‘AttestationKeyType‘ is used by quote3_header (uint16_t att_key_type; ///< 2: sgx_attestation_algorithm_id_t. ), could you set this section name more neutral name?

LGTM，thanks, I will change the `AttestationKeyType` to `SignatureType`.

**Review by zhiminghufighting — APPROVED** [review]

File: rune/libenclave/intelsgx/aesmd.go [hunk 12]
```diff
@@ -100,9 +100,12 @@ func DumpTargetInfo(targetInfo []byte) error {
 	return nil
 }
 
-func DumpLocalReport(report []byte) error {
+/* The report can be used as a separate structure or as a member of the quote structure,
+ * so we add the indent paramter to both support to dump these two cases
```
Comment (jiazhang0) [inline rune/libenclave/intelsgx/aesmd.go:104]: Plz add "." at the end of this line.

File: rune/libenclave/intelsgx/aesmd.go [hunk 13]
```diff
@@ -100,9 +100,12 @@ func DumpTargetInfo(targetInfo []byte) error {
...
+ */
+func DumpLocalReport(report []byte, indent uint32) error {
 	r := &Report{}
-	if err := restruct.Unpack(report, binary.LittleEndian, &r); err != nil {
+	if err := restruct.Unpack(report[indent:indent+ReportLength], binary.LittleEndian, &r); err != nil {
```
Comment (jiazhang0) [inline rune/libenclave/intelsgx/aesmd.go:108]: ????

**Review by jiazhang0 — CHANGES_REQUESTED** [review]

**Review by jiazhang0 — APPROVED** [review]

**Pull request merged**